#Data Wrangling


In [1]:
#uploading data to a path
!wget -O teleCust1000t.csv https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DA0101EN/auto.csv

--2024-07-18 02:06:40--  https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DA0101EN/auto.csv
Resolving s3-api.us-geo.objectstorage.softlayer.net (s3-api.us-geo.objectstorage.softlayer.net)... 67.228.254.196
Connecting to s3-api.us-geo.objectstorage.softlayer.net (s3-api.us-geo.objectstorage.softlayer.net)|67.228.254.196|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 24916 (24K) [text/csv]
Saving to: ‘teleCust1000t.csv’

teleCust1000t.csv   100%[===================>]  24.33K  --.-KB/s    in 0.1s    

2024-07-18 02:06:41 (164 KB/s) - ‘teleCust1000t.csv’ saved [24916/24916]



In [11]:
import pandas as pd
mydata = pd.read_csv('/content/teleCust1000t.csv') #This will read the first row as header

mydata.head(10)

#Notice how the header is wrong

3    ?  alfa-romero  gas    std   two  convertible  rwd  front   88.6  ...  \
0  3    ?  alfa-romero  gas    std   two  convertible  rwd  front   88.6  ...   
1  1    ?  alfa-romero  gas    std   two    hatchback  rwd  front   94.5  ...   
2  2  164         audi  gas    std  four        sedan  fwd  front   99.8  ...   
3  2  164         audi  gas    std  four        sedan  4wd  front   99.4  ...   
4  2    ?         audi  gas    std   two        sedan  fwd  front   99.8  ...   
5  1  158         audi  gas    std  four        sedan  fwd  front  105.8  ...   
6  1    ?         audi  gas    std  four        wagon  fwd  front  105.8  ...   
7  1  158         audi  gas  turbo  four        sedan  fwd  front  105.8  ...   
8  0    ?         audi  gas  turbo   two    hatchback  4wd  front   99.5  ...   
9  2  192          bmw  gas    std   two        sedan  rwd  front  101.2  ...   

   130  mpfi  3.47  2.68   9.0  111  5000  21  27  13495  
0  130  mpfi  3.47  2.68   9.0  111  5000  21  27  16500  
1  152  mpfi  2.68  3.47   9.0  154  5000  19  26  16500  
2  109  mpfi  3.19  3.40  10.0  102  5500  24  30  13950  
3  136  mpfi  3.19  3.40   8.0  115  5500  18  22  17450  
4  136  mpfi  3.19  3.40   8.5  110  5500  19  25  15250  
5  136  mpfi  3.19  3.40   8.5  110  5500  19  25  17710  
6  136  mpfi  3.19  3.40   8.5  110  5500  19  25  18920  
7  131  mpfi  3.13  3.40   8.3  140  5500  17  20  23875  
8  131  mpfi  3.13  3.40   7.0  160  5500  16  22      ?  
9  108  mpfi  3.50  2.80   8.8  101  5800  23  29  16430  

[10 rows x 26 columns]

ERROR ALERT!!
* This is considered wrong because the first row is our data, not the header
* So we need to add the header
* Now we will learn how to add the header to a dataset

In [8]:
mydata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 204 entries, 0 to 203
Data columns (total 26 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   3            204 non-null    int64  
 1   ?            204 non-null    object 
 2   alfa-romero  204 non-null    object 
 3   gas          204 non-null    object 
 4   std          204 non-null    object 
 5   two          204 non-null    object 
 6   convertible  204 non-null    object 
 7   rwd          204 non-null    object 
 8   front        204 non-null    object 
 9   88.6         204 non-null    float64
 10  168.8        204 non-null    float64
 11  64.1         204 non-null    float64
 12  48.8         204 non-null    float64
 13  2548         204 non-null    int64  
 14  dohc         204 non-null    object 
 15  four         204 non-null    object 
 16  130          204 non-null    int64  
 17  mpfi         204 non-null    object 
 18  3.47         204 non-null    object 
 19  2.68    

In [9]:
#check the name of the header columns
mydata.columns

#notice how it gave us the first row of our data

Index(['3', '?', 'alfa-romero', 'gas', 'std', 'two', 'convertible', 'rwd',
       'front', '88.6', '168.8', '64.1', '48.8', '2548', 'dohc', 'four', '130',
       'mpfi', '3.47', '2.68', '9.0', '111', '5000', '21', '27', '13495'],
      dtype='object')

In [12]:
import pandas as pd
mydata = pd.read_csv('/content/teleCust1000t.csv', header = None) #This parameter will not read the first row as the header

mydata.head(10)

#now we have index numbering as the header (good)

0    1            2    3      4     5            6    7      8      9   \
0   3    ?  alfa-romero  gas    std   two  convertible  rwd  front   88.6   
1   3    ?  alfa-romero  gas    std   two  convertible  rwd  front   88.6   
2   1    ?  alfa-romero  gas    std   two    hatchback  rwd  front   94.5   
3   2  164         audi  gas    std  four        sedan  fwd  front   99.8   
4   2  164         audi  gas    std  four        sedan  4wd  front   99.4   
5   2    ?         audi  gas    std   two        sedan  fwd  front   99.8   
6   1  158         audi  gas    std  four        sedan  fwd  front  105.8   
7   1    ?         audi  gas    std  four        wagon  fwd  front  105.8   
8   1  158         audi  gas  turbo  four        sedan  fwd  front  105.8   
9   0    ?         audi  gas  turbo   two    hatchback  4wd  front   99.5   

   ...   16    17    18    19    20   21    22  23  24     25  
0  ...  130  mpfi  3.47  2.68   9.0  111  5000  21  27  13495  
1  ...  130  mpfi  3.47  2.68   9.0  111  5000  21  27  16500  
2  ...  152  mpfi  2.68  3.47   9.0  154  5000  19  26  16500  
3  ...  109  mpfi  3.19  3.40  10.0  102  5500  24  30  13950  
4  ...  136  mpfi  3.19  3.40   8.0  115  5500  18  22  17450  
5  ...  136  mpfi  3.19  3.40   8.5  110  5500  19  25  15250  
6  ...  136  mpfi  3.19  3.40   8.5  110  5500  19  25  17710  
7  ...  136  mpfi  3.19  3.40   8.5  110  5500  19  25  18920  
8  ...  131  mpfi  3.13  3.40   8.3  140  5500  17  20  23875  
9  ...  131  mpfi  3.13  3.40   7.0  160  5500  16  22      ?  

[10 rows x 26 columns]

In [14]:
# Add new headers in columns
name_col = ["symboling","normalized-losses","make","fuel-type","aspiration", "num-of-doors","body-style",
         "drive-wheels","engine-location","wheel-base", "length","width","height","curb-weight","engine-type",
         "num-of-cylinders", "engine-size","fuel-system","bore","stroke","compression-ratio","horsepower",
         "peak-rpm","city-mpg","highway-mpg","price"]

mydata.columns = name_col
mydata.head()

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3                 ?  alfa-romero       gas        std          two   
1          3                 ?  alfa-romero       gas        std          two   
2          1                 ?  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]

In [17]:
#Check the number of rows (No missing data)
mydata.shape

(205, 26)

**Now we see a comparison if we don't use the (header = None) parameter in this particular situation**

In [18]:
import pandas as pd
mydata1 = pd.read_csv('/content/teleCust1000t.csv') #This will read (promote) the first row as header

mydata1.head(10)

#Notice how the header is wrong

3    ?  alfa-romero  gas    std   two  convertible  rwd  front   88.6  ...  \
0  3    ?  alfa-romero  gas    std   two  convertible  rwd  front   88.6  ...   
1  1    ?  alfa-romero  gas    std   two    hatchback  rwd  front   94.5  ...   
2  2  164         audi  gas    std  four        sedan  fwd  front   99.8  ...   
3  2  164         audi  gas    std  four        sedan  4wd  front   99.4  ...   
4  2    ?         audi  gas    std   two        sedan  fwd  front   99.8  ...   
5  1  158         audi  gas    std  four        sedan  fwd  front  105.8  ...   
6  1    ?         audi  gas    std  four        wagon  fwd  front  105.8  ...   
7  1  158         audi  gas  turbo  four        sedan  fwd  front  105.8  ...   
8  0    ?         audi  gas  turbo   two    hatchback  4wd  front   99.5  ...   
9  2  192          bmw  gas    std   two        sedan  rwd  front  101.2  ...   

   130  mpfi  3.47  2.68   9.0  111  5000  21  27  13495  
0  130  mpfi  3.47  2.68   9.0  111  5000  21  27  16500  
1  152  mpfi  2.68  3.47   9.0  154  5000  19  26  16500  
2  109  mpfi  3.19  3.40  10.0  102  5500  24  30  13950  
3  136  mpfi  3.19  3.40   8.0  115  5500  18  22  17450  
4  136  mpfi  3.19  3.40   8.5  110  5500  19  25  15250  
5  136  mpfi  3.19  3.40   8.5  110  5500  19  25  17710  
6  136  mpfi  3.19  3.40   8.5  110  5500  19  25  18920  
7  131  mpfi  3.13  3.40   8.3  140  5500  17  20  23875  
8  131  mpfi  3.13  3.40   7.0  160  5500  16  22      ?  
9  108  mpfi  3.50  2.80   8.8  101  5800  23  29  16430  

[10 rows x 26 columns]

In [19]:
# Add new headers in columns
name_col = ["symboling","normalized-losses","make","fuel-type","aspiration", "num-of-doors","body-style",
         "drive-wheels","engine-location","wheel-base", "length","width","height","curb-weight","engine-type",
         "num-of-cylinders", "engine-size","fuel-system","bore","stroke","compression-ratio","horsepower",
         "peak-rpm","city-mpg","highway-mpg","price"]

mydata1.columns = name_col
mydata1.head()

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3                 ?  alfa-romero       gas        std          two   
1          1                 ?  alfa-romero       gas        std          two   
2          2               164         audi       gas        std         four   
3          2               164         audi       gas        std         four   
4          2                 ?         audi       gas        std          two   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1    hatchback          rwd           front        94.5  ...          152   
2        sedan          fwd           front        99.8  ...          109   
3        sedan          4wd           front        99.4  ...          136   
4        sedan          fwd           front        99.8  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  2.68    3.47               9.0        154      5000       19   
2         mpfi  3.19    3.40              10.0        102      5500       24   
3         mpfi  3.19    3.40               8.0        115      5500       18   
4         mpfi  3.19    3.40               8.5        110      5500       19   

  highway-mpg  price  
0          27  16500  
1          26  16500  
2          30  13950  
3          22  17450  
4          25  15250  

[5 rows x 26 columns]

In [20]:
mydata1.shape

(204, 26)

In [23]:
mydata1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 204 entries, 0 to 203
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          204 non-null    int64  
 1   normalized-losses  204 non-null    object 
 2   make               204 non-null    object 
 3   fuel-type          204 non-null    object 
 4   aspiration         204 non-null    object 
 5   num-of-doors       204 non-null    object 
 6   body-style         204 non-null    object 
 7   drive-wheels       204 non-null    object 
 8   engine-location    204 non-null    object 
 9   wheel-base         204 non-null    float64
 10  length             204 non-null    float64
 11  width              204 non-null    float64
 12  height             204 non-null    float64
 13  curb-weight        204 non-null    int64  
 14  engine-type        204 non-null    object 
 15  num-of-cylinders   204 non-null    object 
 16  engine-size        204 non

**Missing Value**

* read each row to find missing data
* normalized losses column is supposed to be numerical data but pyhton considered it as categorical data (so this is wrong)

In [25]:
mydata.describe(include='all')

# read each row to find missing data
# normalized losses column is supposed to be numerical data but pyhton considered it as categorical data (so this is wrong)

symboling normalized-losses    make fuel-type aspiration  \
count   205.000000               205     205       205        205   
unique         NaN                52      22         2          2   
top            NaN                 ?  toyota       gas        std   
freq           NaN                41      32       185        168   
mean      0.834146               NaN     NaN       NaN        NaN   
std       1.245307               NaN     NaN       NaN        NaN   
min      -2.000000               NaN     NaN       NaN        NaN   
25%       0.000000               NaN     NaN       NaN        NaN   
50%       1.000000               NaN     NaN       NaN        NaN   
75%       2.000000               NaN     NaN       NaN        NaN   
max       3.000000               NaN     NaN       NaN        NaN   

       num-of-doors body-style drive-wheels engine-location  wheel-base  ...  \
count           205        205          205             205  205.000000  ...   
unique            3          5            3               2         NaN  ...   
top            four      sedan          fwd           front         NaN  ...   
freq            114         96          120             202         NaN  ...   
mean            NaN        NaN          NaN             NaN   98.756585  ...   
std             NaN        NaN          NaN             NaN    6.021776  ...   
min             NaN        NaN          NaN             NaN   86.600000  ...   
25%             NaN        NaN          NaN             NaN   94.500000  ...   
50%             NaN        NaN          NaN             NaN   97.000000  ...   
75%             NaN        NaN          NaN             NaN  102.400000  ...   
max             NaN        NaN          NaN             NaN  120.900000  ...   

        engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
count    205.000000          205   205     205        205.000000        205   
unique          NaN            8    39      37               NaN         60   
top             NaN         mpfi  3.62    3.40               NaN         68   
freq            NaN           94    23      20               NaN         19   
mean     126.907317          NaN   NaN     NaN         10.142537        NaN   
std       41.642693          NaN   NaN     NaN          3.972040        NaN   
min       61.000000          NaN   NaN     NaN          7.000000        NaN   
25%       97.000000          NaN   NaN     NaN          8.600000        NaN   
50%      120.000000          NaN   NaN     NaN          9.000000        NaN   
75%      141.000000          NaN   NaN     NaN          9.400000        NaN   
max      326.000000          NaN   NaN     NaN         23.000000        NaN   

        peak-rpm    city-mpg highway-mpg price  
count        205  205.000000  205.000000   205  
unique        24         NaN         NaN   187  
top         5500         NaN         NaN     ?  
freq          37         NaN         NaN     4  
mean         NaN   25.219512   30.751220   NaN  
std          NaN    6.542142    6.886443   NaN  
min          NaN   13.000000   16.000000   NaN  
25%          NaN   19.000000   25.000000   NaN  
50%          NaN   24.000000   30.000000   NaN  
75%          NaN   30.000000   34.000000   NaN  
max          NaN   49.000000   54.000000   NaN  

[11 rows x 26 columns]

* Display all the columns for us to check each columns if it has a missing value

In [26]:
# Display all the columns for us to check each columns if it has a missing value

pd.set_option('display.max_columns', None)
mydata.head(10)

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3                 ?  alfa-romero       gas        std          two   
1          3                 ?  alfa-romero       gas        std          two   
2          1                 ?  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   
5          2                 ?         audi       gas        std          two   
6          1               158         audi       gas        std         four   
7          1                 ?         audi       gas        std         four   
8          1               158         audi       gas      turbo         four   
9          0                 ?         audi       gas      turbo          two   

    body-style drive-wheels engine-location  wheel-base  length  width  \
0  convertible          rwd           front        88.6   168.8   64.1   
1  convertible          rwd           front        88.6   168.8   64.1   
2    hatchback          rwd           front        94.5   171.2   65.5   
3        sedan          fwd           front        99.8   176.6   66.2   
4        sedan          4wd           front        99.4   176.6   66.4   
5        sedan          fwd           front        99.8   177.3   66.3   
6        sedan          fwd           front       105.8   192.7   71.4   
7        wagon          fwd           front       105.8   192.7   71.4   
8        sedan          fwd           front       105.8   192.7   71.4   
9    hatchback          4wd           front        99.5   178.2   67.9   

   height  curb-weight engine-type num-of-cylinders  engine-size fuel-system  \
0    48.8         2548        dohc             four          130        mpfi   
1    48.8         2548        dohc             four          130        mpfi   
2    52.4         2823        ohcv              six          152        mpfi   
3    54.3         2337         ohc             four          109        mpfi   
4    54.3         2824         ohc             five          136        mpfi   
5    53.1         2507         ohc             five          136        mpfi   
6    55.7         2844         ohc             five          136        mpfi   
7    55.7         2954         ohc             five          136        mpfi   
8    55.9         3086         ohc             five          131        mpfi   
9    52.0         3053         ohc             five          131        mpfi   

   bore stroke  compression-ratio horsepower peak-rpm  city-mpg  highway-mpg  \
0  3.47   2.68                9.0        111     5000        21           27   
1  3.47   2.68                9.0        111     5000        21           27   
2  2.68   3.47                9.0        154     5000        19           26   
3  3.19   3.40               10.0        102     5500        24           30   
4  3.19   3.40                8.0        115     5500        18           22   
5  3.19   3.40                8.5        110     5500        19           25   
6  3.19   3.40                8.5        110     5500        19           25   
7  3.19   3.40                8.5        110     5500        19           25   
8  3.13   3.40                8.3        140     5500        17           20   
9  3.13   3.40                7.0        160     5500        16           22   

   price  
0  13495  
1  16500  
2  16500  
3  13950  
4  17450  
5  15250  
6  17710  
7  18920  
8  23875  
9      ?

**Checking missing value has 2 methods**

**Method 1:**

In [29]:
#Check missing value
#Method 1:

mydata.isnull()

#False = No missing value
#True = There's a missing value

symboling  normalized-losses   make  fuel-type  aspiration  num-of-doors  \
0        False              False  False      False       False         False   
1        False              False  False      False       False         False   
2        False              False  False      False       False         False   
3        False              False  False      False       False         False   
4        False              False  False      False       False         False   
..         ...                ...    ...        ...         ...           ...   
200      False              False  False      False       False         False   
201      False              False  False      False       False         False   
202      False              False  False      False       False         False   
203      False              False  False      False       False         False   
204      False              False  False      False       False         False   

     body-style  drive-wheels  engine-location  wheel-base  length  width  \
0         False         False            False       False   False  False   
1         False         False            False       False   False  False   
2         False         False            False       False   False  False   
3         False         False            False       False   False  False   
4         False         False            False       False   False  False   
..          ...           ...              ...         ...     ...    ...   
200       False         False            False       False   False  False   
201       False         False            False       False   False  False   
202       False         False            False       False   False  False   
203       False         False            False       False   False  False   
204       False         False            False       False   False  False   

     height  curb-weight  engine-type  num-of-cylinders  engine-size  \
0     False        False        False             False        False   
1     False        False        False             False        False   
2     False        False        False             False        False   
3     False        False        False             False        False   
4     False        False        False             False        False   
..      ...          ...          ...               ...          ...   
200   False        False        False             False        False   
201   False        False        False             False        False   
202   False        False        False             False        False   
203   False        False        False             False        False   
204   False        False        False             False        False   

     fuel-system   bore  stroke  compression-ratio  horsepower  peak-rpm  \
0          False  False   False              False       False     False   
1          False  False   False              False       False     False   
2          False  False   False              False       False     False   
3          False  False   False              False       False     False   
4          False  False   False              False       False     False   
..           ...    ...     ...                ...         ...       ...   
200        False  False   False              False       False     False   
201        False  False   False              False       False     False   
202        False  False   False              False       False     False   
203        False  False   False              False       False     False   
204        False  False   False              False       False     False   

     city-mpg  highway-mpg  price  
0       False        False  False  
1       False        False  False  
2       False        False  False  
3       False        False  False  
4       False        False  False  
..        ...          ...    ...  
200     False        False  False  
201     False        False  False  
202  

**Method 2:**

In [28]:
#Method 2:
#Checking total number of missing value in a column
mydata.isnull().sum()

symboling            0
normalized-losses    0
make                 0
fuel-type            0
aspiration           0
num-of-doors         0
body-style           0
drive-wheels         0
engine-location      0
wheel-base           0
length               0
width                0
height               0
curb-weight          0
engine-type          0
num-of-cylinders     0
engine-size          0
fuel-system          0
bore                 0
stroke               0
compression-ratio    0
horsepower           0
peak-rpm             0
city-mpg             0
highway-mpg          0
price                0
dtype: int64

**ERROR ALER!!**
* Notice how the result are all zero (0) eventhough we are sure there is a missing value in our data
* The missing data in our data appeared as question mark (?)
* So question mark (?) in python is considered as a string
* So numpy library read the data string as object

**Replace the missing value '?' to NaN numpy**

In [30]:
#replace the '?' by NaN numpy
import numpy as np

mydata.replace('?', np.nan, inplace = True)

#Checking total number of missing value in a column AGAIN
mydata.isnull().sum()

symboling             0
normalized-losses    41
make                  0
fuel-type             0
aspiration            0
num-of-doors          2
body-style            0
drive-wheels          0
engine-location       0
wheel-base            0
length                0
width                 0
height                0
curb-weight           0
engine-type           0
num-of-cylinders      0
engine-size           0
fuel-system           0
bore                  4
stroke                4
compression-ratio     0
horsepower            2
peak-rpm              2
city-mpg              0
highway-mpg           0
price                 4
dtype: int64

**Note:**
* Notice how the missing value appears as we replace '?' by NaN numpy

In [31]:
mydata.head()

#Notice how the 'normalized losses' missing value is now replaced with NaN numpy

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3               NaN  alfa-romero       gas        std          two   
1          3               NaN  alfa-romero       gas        std          two   
2          1               NaN  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  length  width  \
0  convertible          rwd           front        88.6   168.8   64.1   
1  convertible          rwd           front        88.6   168.8   64.1   
2    hatchback          rwd           front        94.5   171.2   65.5   
3        sedan          fwd           front        99.8   176.6   66.2   
4        sedan          4wd           front        99.4   176.6   66.4   

   height  curb-weight engine-type num-of-cylinders  engine-size fuel-system  \
0    48.8         2548        dohc             four          130        mpfi   
1    48.8         2548        dohc             four          130        mpfi   
2    52.4         2823        ohcv              six          152        mpfi   
3    54.3         2337         ohc             four          109        mpfi   
4    54.3         2824         ohc             five          136        mpfi   

   bore stroke  compression-ratio horsepower peak-rpm  city-mpg  highway-mpg  \
0  3.47   2.68                9.0        111     5000        21           27   
1  3.47   2.68                9.0        111     5000        21           27   
2  2.68   3.47                9.0        154     5000        19           26   
3  3.19   3.40               10.0        102     5500        24           30   
4  3.19   3.40                8.0        115     5500        18           22   

   price  
0  13495  
1  16500  
2  16500  
3  13950  
4  17450

In [32]:
mydata.info()

#Notice how 'normalized-losses' has 164 value.
#So total data in a column 205 - 164 = 41 (total of missing value in a column)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          205 non-null    int64  
 1   normalized-losses  164 non-null    object 
 2   make               205 non-null    object 
 3   fuel-type          205 non-null    object 
 4   aspiration         205 non-null    object 
 5   num-of-doors       203 non-null    object 
 6   body-style         205 non-null    object 
 7   drive-wheels       205 non-null    object 
 8   engine-location    205 non-null    object 
 9   wheel-base         205 non-null    float64
 10  length             205 non-null    float64
 11  width              205 non-null    float64
 12  height             205 non-null    float64
 13  curb-weight        205 non-null    int64  
 14  engine-type        205 non-null    object 
 15  num-of-cylinders   205 non-null    object 
 16  engine-size        205 non

Columns with missing values:

<ol>
    <li>"normalized-losses": 41 missing data (change with the mean)</li>
    <li>"num-of-doors": 2 missing data (change to most frequent categorical)</li>
    <li>"bore": 4 missing data (replace with mean)</li>
    <li>"stroke" : 4 missing data (replace with mean)</li>
    <li>"horsepower": 2 missing data (replace with mean)</li>
    <li>"peak-rpm": 2 missing data (replace with mean)</li>
    <li>"price": 4 missing data (replace with mean)</li>
</ol>

## Dealing With Missing Value In "Normalized-Losses"

In [36]:
mean_norm = mydata['normalized-losses'].mean()

mean_norm

TypeError: unsupported operand type(s) for +: 'int' and 'str'

**ERROR ALERT!!**

* Remember that the 'normalized-loses' is an "**object**" so **CONVERT** it to "**float**"

**DANGER!!!**
* If you convert float to integer, you will lose some data because it will round up the decimal number to a whole number
* But if you convert interger to float then it will not matter

In [37]:
mean_norm = mydata['normalized-losses'].astype(float).mean()

mean_norm

122.0

**Replace Function**
* replace(value nak replace, value baru, inplace = True)

In [39]:
mydata['normalized-losses'].replace(np.nan, mean_norm, inplace = True)

#Checking total number of missing value in a column AGAIN
mydata.isnull().sum()

symboling            0
normalized-losses    0
make                 0
fuel-type            0
aspiration           0
num-of-doors         2
body-style           0
drive-wheels         0
engine-location      0
wheel-base           0
length               0
width                0
height               0
curb-weight          0
engine-type          0
num-of-cylinders     0
engine-size          0
fuel-system          0
bore                 4
stroke               4
compression-ratio    0
horsepower           2
peak-rpm             2
city-mpg             0
highway-mpg          0
price                4
dtype: int64

## Dealing With Missing Value In "num-of-doors"

### 3 Methods To Check Frequency of A Columns

**Method 1: Using Describe Function**

In [43]:
#Cange number of doors missing value with the most frequency categorical value

#check frequency of the categorical value on number of doors column
mydata.describe(include='all')

symboling  normalized-losses    make fuel-type aspiration  \
count   205.000000              205.0     205       205        205   
unique         NaN               52.0      22         2          2   
top            NaN              122.0  toyota       gas        std   
freq           NaN               41.0      32       185        168   
mean      0.834146                NaN     NaN       NaN        NaN   
std       1.245307                NaN     NaN       NaN        NaN   
min      -2.000000                NaN     NaN       NaN        NaN   
25%       0.000000                NaN     NaN       NaN        NaN   
50%       1.000000                NaN     NaN       NaN        NaN   
75%       2.000000                NaN     NaN       NaN        NaN   
max       3.000000                NaN     NaN       NaN        NaN   

       num-of-doors body-style drive-wheels engine-location  wheel-base  \
count           203        205          205             205  205.000000   
unique            2          5            3               2         NaN   
top            four      sedan          fwd           front         NaN   
freq            114         96          120             202         NaN   
mean            NaN        NaN          NaN             NaN   98.756585   
std             NaN        NaN          NaN             NaN    6.021776   
min             NaN        NaN          NaN             NaN   86.600000   
25%             NaN        NaN          NaN             NaN   94.500000   
50%             NaN        NaN          NaN             NaN   97.000000   
75%             NaN        NaN          NaN             NaN  102.400000   
max             NaN        NaN          NaN             NaN  120.900000   

            length       width      height  curb-weight engine-type  \
count   205.000000  205.000000  205.000000   205.000000         205   
unique         NaN         NaN         NaN          NaN           7   
top            NaN         NaN         NaN          NaN         ohc   
freq           NaN         NaN         NaN          NaN         148   
mean    174.049268   65.907805   53.724878  2555.565854         NaN   
std      12.337289    2.145204    2.443522   520.680204         NaN   
min     141.100000   60.300000   47.800000  1488.000000         NaN   
25%     166.300000   64.100000   52.000000  2145.000000         NaN   
50%     173.200000   65.500000   54.100000  2414.000000         NaN   
75%     183.100000   66.900000   55.500000  2935.000000         NaN   
max     208.100000   72.300000   59.800000  4066.000000         NaN   

       num-of-cylinders  engine-size fuel-system  bore stroke  \
count               205   205.000000         205   201    201   
unique                7          NaN           8    38     36   
top                four          NaN        mpfi  3.62   3.40   
freq                159          NaN          94    23     20   
mean                NaN   126.907317         NaN   NaN    NaN   
std                 NaN    41.642693         NaN   NaN    NaN   
min                 NaN    61.000000         NaN   NaN    NaN   
25%                 NaN    97.000000         NaN   NaN    NaN   
50%                 NaN   120.000000         NaN   NaN    NaN   
75%                 NaN   141.000000         NaN   NaN    NaN   
max                 NaN   326.000000         NaN   NaN    NaN   

        compression-ratio horsepower peak-rpm    city-mpg  highway-mpg price  
count          205.000000        203      203  205.000000   205.000000   201  
unique                NaN         59       23         NaN          NaN   186  
top                   NaN         68     5500         NaN          NaN  8921  
freq                  NaN         19       37         NaN          NaN     2  
mean            10.142537        NaN      NaN   25.219512    30.751220   NaN  
std              3.972040        NaN      NaN    6.542142     6.886443   NaN  
min              7.000000        NaN      NaN   13.000000    16.000000   NaN  
2

**Method 2: Using Count Function**

In [41]:
mydata['num-of-doors'].value_counts()

num-of-doors
four    114
two      89
Name: count, dtype: int64

**Method 3: Using Count Max Value Function**

In [46]:
mydata['num-of-doors'].value_counts().idxmax()

'four'

* replace the "num-of-doors" missing value with 'four' the most frequent value in that column

In [47]:
#Cara 1:
mydata['num-of-doors'].replace(np.nan,mydata['num-of-doors'].value_counts().idxmax(), inplace=True)

#Cara 2:
#mydata['num-of-doors'].replace(np.nan,'four', inplace=True)
#cara 2 pon betul (mana-mana pon boleh)

mydata.isnull().sum()

symboling            0
normalized-losses    0
make                 0
fuel-type            0
aspiration           0
num-of-doors         0
body-style           0
drive-wheels         0
engine-location      0
wheel-base           0
length               0
width                0
height               0
curb-weight          0
engine-type          0
num-of-cylinders     0
engine-size          0
fuel-system          0
bore                 4
stroke               4
compression-ratio    0
horsepower           2
peak-rpm             2
city-mpg             0
highway-mpg          0
price                4
dtype: int64

## Dealing With Missing Value In "bore"

In [48]:
mean_bore= mydata['bore'].astype(float).mean()

mean_bore

3.3297512437810943

In [49]:
mydata['bore'].replace(np.nan, mean_bore, inplace = True)

#Checking total number of missing value the bore column AGAIN
mydata.isnull().sum()

symboling            0
normalized-losses    0
make                 0
fuel-type            0
aspiration           0
num-of-doors         0
body-style           0
drive-wheels         0
engine-location      0
wheel-base           0
length               0
width                0
height               0
curb-weight          0
engine-type          0
num-of-cylinders     0
engine-size          0
fuel-system          0
bore                 0
stroke               4
compression-ratio    0
horsepower           2
peak-rpm             2
city-mpg             0
highway-mpg          0
price                4
dtype: int64

## Dealing With The Rest of The Missing Value

In [53]:
# Finding the mean value of the 'stroke' column
mean_stroke= mydata['stroke'].astype(float).mean()
mean_horsepower= mydata['horsepower'].astype(float).mean()
mean_peak= mydata['peak-rpm'].astype(float).mean()

# Replace the missing value in 'stroke' with the mean value
mydata['stroke'].replace(np.nan, mean_stroke, inplace = True)
mydata['horsepower'].replace(np.nan, mean_horsepower, inplace = True)
mydata['peak-rpm'].replace(np.nan, mean_peak, inplace = True)

#Checking total number of missing value the stroke column AGAIN
mydata.isnull().sum()

symboling            0
normalized-losses    0
make                 0
fuel-type            0
aspiration           0
num-of-doors         0
body-style           0
drive-wheels         0
engine-location      0
wheel-base           0
length               0
width                0
height               0
curb-weight          0
engine-type          0
num-of-cylinders     0
engine-size          0
fuel-system          0
bore                 0
stroke               0
compression-ratio    0
horsepower           0
peak-rpm             0
city-mpg             0
highway-mpg          0
price                4
dtype: int64

## Dealing With Missing Value In "Price"
* Here we are going to drop the missing value in price

In [55]:
mydata.dropna(subset=['price'],axis=0, inplace=True)

#Checking total number of missing value the price column AGAIN
mydata.isnull().sum()

symboling            0
normalized-losses    0
make                 0
fuel-type            0
aspiration           0
num-of-doors         0
body-style           0
drive-wheels         0
engine-location      0
wheel-base           0
length               0
width                0
height               0
curb-weight          0
engine-type          0
num-of-cylinders     0
engine-size          0
fuel-system          0
bore                 0
stroke               0
compression-ratio    0
horsepower           0
peak-rpm             0
city-mpg             0
highway-mpg          0
price                0
dtype: int64

* Now we check the total number of rows
* Remaining 201, because we remove 4 rows from the missing price value

In [56]:
mydata.shape

(201, 26)

**ALERT!!**

* Why is the datatype didn't change?
  * Answer: This is because we only change the datatype for the missing value NaN not the whole data

In [57]:
mydata.info()

<class 'pandas.core.frame.DataFrame'>
Index: 201 entries, 0 to 204
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          201 non-null    int64  
 1   normalized-losses  201 non-null    object 
 2   make               201 non-null    object 
 3   fuel-type          201 non-null    object 
 4   aspiration         201 non-null    object 
 5   num-of-doors       201 non-null    object 
 6   body-style         201 non-null    object 
 7   drive-wheels       201 non-null    object 
 8   engine-location    201 non-null    object 
 9   wheel-base         201 non-null    float64
 10  length             201 non-null    float64
 11  width              201 non-null    float64
 12  height             201 non-null    float64
 13  curb-weight        201 non-null    int64  
 14  engine-type        201 non-null    object 
 15  num-of-cylinders   201 non-null    object 
 16  engine-size        201 non-null

In [60]:
# 2 Methods to change the datatype

#Tulis satu-satu
mydata['normalized-losses'] = mydata['normalized-losses'].astype('float')

#Group it all (don't forgot double square bracket [[]] if you want to group it)
mydata[['bore','stroke','horsepower','peak-rpm','price']] = mydata[['bore','stroke','horsepower','peak-rpm','price']].astype('float')

mydata.dtypes

symboling              int64
normalized-losses    float64
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                 float64
stroke               float64
compression-ratio    float64
horsepower           float64
peak-rpm             float64
city-mpg               int64
highway-mpg            int64
price                float64
dtype: object

In [62]:
mydata.head()

#mpg = mile per galon CONVERT to Kilometer

symboling  normalized-losses         make fuel-type aspiration  \
0          3              122.0  alfa-romero       gas        std   
1          3              122.0  alfa-romero       gas        std   
2          1              122.0  alfa-romero       gas        std   
3          2              164.0         audi       gas        std   
4          2              164.0         audi       gas        std   

  num-of-doors   body-style drive-wheels engine-location  wheel-base  length  \
0          two  convertible          rwd           front        88.6   168.8   
1          two  convertible          rwd           front        88.6   168.8   
2          two    hatchback          rwd           front        94.5   171.2   
3         four        sedan          fwd           front        99.8   176.6   
4         four        sedan          4wd           front        99.4   176.6   

   width  height  curb-weight engine-type num-of-cylinders  engine-size  \
0   64.1    48.8         2548        dohc             four          130   
1   64.1    48.8         2548        dohc             four          130   
2   65.5    52.4         2823        ohcv              six          152   
3   66.2    54.3         2337         ohc             four          109   
4   66.4    54.3         2824         ohc             five          136   

  fuel-system  bore  stroke  compression-ratio  horsepower  peak-rpm  \
0        mpfi  3.47    2.68                9.0       111.0    5000.0   
1        mpfi  3.47    2.68                9.0       111.0    5000.0   
2        mpfi  2.68    3.47                9.0       154.0    5000.0   
3        mpfi  3.19    3.40               10.0       102.0    5500.0   
4        mpfi  3.19    3.40                8.0       115.0    5500.0   

   city-mpg  highway-mpg    price  
0        21           27  13495.0  
1        21           27  16500.0  
2        19           26  16500.0  
3        24           30  13950.0  
4        18           22  17450.0

In [63]:
mydata['city_L/100KM'] = 235/mydata['city-mpg']
mydata['highway_L/100KM'] = 235/mydata['highway-mpg']

mydata.head()

symboling  normalized-losses         make fuel-type aspiration  \
0          3              122.0  alfa-romero       gas        std   
1          3              122.0  alfa-romero       gas        std   
2          1              122.0  alfa-romero       gas        std   
3          2              164.0         audi       gas        std   
4          2              164.0         audi       gas        std   

  num-of-doors   body-style drive-wheels engine-location  wheel-base  length  \
0          two  convertible          rwd           front        88.6   168.8   
1          two  convertible          rwd           front        88.6   168.8   
2          two    hatchback          rwd           front        94.5   171.2   
3         four        sedan          fwd           front        99.8   176.6   
4         four        sedan          4wd           front        99.4   176.6   

   width  height  curb-weight engine-type num-of-cylinders  engine-size  \
0   64.1    48.8         2548        dohc             four          130   
1   64.1    48.8         2548        dohc             four          130   
2   65.5    52.4         2823        ohcv              six          152   
3   66.2    54.3         2337         ohc             four          109   
4   66.4    54.3         2824         ohc             five          136   

  fuel-system  bore  stroke  compression-ratio  horsepower  peak-rpm  \
0        mpfi  3.47    2.68                9.0       111.0    5000.0   
1        mpfi  3.47    2.68                9.0       111.0    5000.0   
2        mpfi  2.68    3.47                9.0       154.0    5000.0   
3        mpfi  3.19    3.40               10.0       102.0    5500.0   
4        mpfi  3.19    3.40                8.0       115.0    5500.0   

   city-mpg  highway-mpg    price  city_L/100KM  highway_L/100KM  
0        21           27  13495.0     11.190476         8.703704  
1        21           27  16500.0     11.190476         8.703704  
2        19           26  16500.0     12.368421         9.038462  
3        24           30  13950.0      9.791667         7.833333  
4        18           22  17450.0     13.055556        10.681818

* Since we already created the convert mpg to KM
* We can now delete the city-mpg and highway-mpg columns

In [64]:
mydata.drop('city-mpg',axis=1, inplace=True)
mydata.drop('highway-mpg',axis=1, inplace=True)

mydata.head()

symboling  normalized-losses         make fuel-type aspiration  \
0          3              122.0  alfa-romero       gas        std   
1          3              122.0  alfa-romero       gas        std   
2          1              122.0  alfa-romero       gas        std   
3          2              164.0         audi       gas        std   
4          2              164.0         audi       gas        std   

  num-of-doors   body-style drive-wheels engine-location  wheel-base  length  \
0          two  convertible          rwd           front        88.6   168.8   
1          two  convertible          rwd           front        88.6   168.8   
2          two    hatchback          rwd           front        94.5   171.2   
3         four        sedan          fwd           front        99.8   176.6   
4         four        sedan          4wd           front        99.4   176.6   

   width  height  curb-weight engine-type num-of-cylinders  engine-size  \
0   64.1    48.8         2548        dohc             four          130   
1   64.1    48.8         2548        dohc             four          130   
2   65.5    52.4         2823        ohcv              six          152   
3   66.2    54.3         2337         ohc             four          109   
4   66.4    54.3         2824         ohc             five          136   

  fuel-system  bore  stroke  compression-ratio  horsepower  peak-rpm    price  \
0        mpfi  3.47    2.68                9.0       111.0    5000.0  13495.0   
1        mpfi  3.47    2.68                9.0       111.0    5000.0  16500.0   
2        mpfi  2.68    3.47                9.0       154.0    5000.0  16500.0   
3        mpfi  3.19    3.40               10.0       102.0    5500.0  13950.0   
4        mpfi  3.19    3.40                8.0       115.0    5500.0  17450.0   

   city_L/100KM  highway_L/100KM  
0     11.190476         8.703704  
1     11.190476         8.703704  
2     12.368421         9.038462  
3      9.791667         7.833333  
4     13.055556        10.681818

**Difference between '.drop' and '.dropna'**

**mydata.drop:**

* Removes rows or columns based on specific labels.
* Can remove specific rows by index or specific columns by name.
* Useful for removing specific parts of the DataFrame.

**mydata.dropna:**

* Removes rows or columns with missing values.
* Can remove rows or columns with any or all missing values.
* Useful for cleaning data by removing incomplete rows or columns.

## Data Normalization

In [65]:
#Checking the width data
mydata['width'].head()

0    64.1
1    64.1
2    65.5
3    66.2
4    66.4
Name: width, dtype: float64

In [73]:
#Checking the width, length, height data
mydata[['width','length','height']].head()

width    length  height
0  0.890278  0.413433    48.8
1  0.890278  0.413433    48.8
2  0.909722  0.449254    52.4
3  0.919444  0.529851    54.3
4  0.922222  0.529851    54.3

### Data Normalization: Using Simple Feature Scaling

In [76]:
#Simple Feature Scaling for the width column
mydata['width'] = mydata['width']/mydata['width'].max()
mydata[['width','length','height']].head()

width    length    height
0  0.890278  0.413433 -2.029015
1  0.890278  0.413433 -2.029015
2  0.909722  0.449254 -0.558319
3  0.919444  0.529851  0.217881
4  0.922222  0.529851  0.217881

In [71]:
mydata['width'].describe()

count    201.000000
mean       0.915126
std        0.029187
min        0.837500
25%        0.890278
50%        0.909722
75%        0.925000
max        1.000000
Name: width, dtype: float64

### Data Normalization: Using Min-Max Method

In [72]:
#min-max normalization for length column
mydata['length'] = (mydata['length']-mydata['length'].min())/(mydata['length'].max()-mydata['length'].min())
mydata[['width','length','height']].head()

width    length  height
0  0.890278  0.413433    48.8
1  0.890278  0.413433    48.8
2  0.909722  0.449254    52.4
3  0.919444  0.529851    54.3
4  0.922222  0.529851    54.3

### Data Normalization: Using Z-Score Method

In [74]:
#z-score normalization for height column
mydata['height'] = (mydata['height']-mydata['height'].mean())/mydata['height'].std()
mydata[['width','length','height']].head()

width    length    height
0  0.890278  0.413433 -2.029015
1  0.890278  0.413433 -2.029015
2  0.909722  0.449254 -0.558319
3  0.919444  0.529851  0.217881
4  0.922222  0.529851  0.217881

In [77]:
#Now let's compare all the data normalization
#Usually we don't use all three methods at once
#It is recommended to only use one data normalization method throughout the project
#People usually use the z-score method
mydata[['width','length','height']].describe()

width      length        height
count  201.000000  201.000000  2.010000e+02
mean     0.915126    0.494045  4.949054e-16
std      0.029187    0.183913  1.000000e+00
min      0.837500    0.000000 -2.437541e+00
25%      0.890278    0.383582 -7.217300e-01
50%      0.909722    0.479104  1.361755e-01
75%      0.925000    0.632836  7.081124e-01
max      1.000000    1.000000  2.464776e+00

**Converting categorical to numerical**

* one hot-encoding
* encoder
* other

In [78]:
mydata.describe(include='all')

symboling  normalized-losses    make fuel-type aspiration  \
count   201.000000          201.00000     201       201        201   
unique         NaN                NaN      22         2          2   
top            NaN                NaN  toyota       gas        std   
freq           NaN                NaN      32       181        165   
mean      0.840796          122.00000     NaN       NaN        NaN   
std       1.254802           31.99625     NaN       NaN        NaN   
min      -2.000000           65.00000     NaN       NaN        NaN   
25%       0.000000          101.00000     NaN       NaN        NaN   
50%       1.000000          122.00000     NaN       NaN        NaN   
75%       2.000000          137.00000     NaN       NaN        NaN   
max       3.000000          256.00000     NaN       NaN        NaN   

       num-of-doors body-style drive-wheels engine-location  wheel-base  \
count           201        201          201             201  201.000000   
unique            2          5            3               2         NaN   
top            four      sedan          fwd           front         NaN   
freq            115         94          118             198         NaN   
mean            NaN        NaN          NaN             NaN   98.797015   
std             NaN        NaN          NaN             NaN    6.066366   
min             NaN        NaN          NaN             NaN   86.600000   
25%             NaN        NaN          NaN             NaN   94.500000   
50%             NaN        NaN          NaN             NaN   97.000000   
75%             NaN        NaN          NaN             NaN  102.400000   
max             NaN        NaN          NaN             NaN  120.900000   

            length       width        height  curb-weight engine-type  \
count   201.000000  201.000000  2.010000e+02   201.000000         201   
unique         NaN         NaN           NaN          NaN           6   
top            NaN         NaN           NaN          NaN         ohc   
freq           NaN         NaN           NaN          NaN         145   
mean      0.494045    0.915126  4.949054e-16  2555.666667         NaN   
std       0.183913    0.029187  1.000000e+00   517.296727         NaN   
min       0.000000    0.837500 -2.437541e+00  1488.000000         NaN   
25%       0.383582    0.890278 -7.217300e-01  2169.000000         NaN   
50%       0.479104    0.909722  1.361755e-01  2414.000000         NaN   
75%       0.632836    0.925000  7.081124e-01  2926.000000         NaN   
max       1.000000    1.000000  2.464776e+00  4066.000000         NaN   

       num-of-cylinders  engine-size fuel-system        bore      stroke  \
count               201   201.000000         201  201.000000  201.000000   
unique                7          NaN           8         NaN         NaN   
top                four          NaN        mpfi         NaN         NaN   
freq                157          NaN          92         NaN         NaN   
mean                NaN   126.875622         NaN    3.330692    3.256874   
std                 NaN    41.546834         NaN    0.268072    0.316048   
min                 NaN    61.000000         NaN    2.540000    2.070000   
25%                 NaN    98.000000         NaN    3.150000    3.110000   
50%                 NaN   120.000000         NaN    3.310000    3.290000   
75%                 NaN   141.000000         NaN    3.580000    3.410000   
max                 NaN   326.000000         NaN    3.940000    4.170000   

        compression-ratio  horsepower     peak-rpm         price  \
count          201.000000  201.000000   201.000000    201.000000   
unique                NaN         NaN          NaN           NaN   
top                   NaN         NaN          NaN           NaN   
freq                  NaN         NaN          NaN           NaN   
mean            10.164279  103.405534  5117.665368  13207.129353   
std              4.004965   37.365700   478.113805   7947.066342   
m

In [79]:
mydata[['num-of-doors','num-of-cylinders']].head(20)

num-of-doors num-of-cylinders
0           two             four
1           two             four
2           two              six
3          four             four
4          four             five
5           two             five
6          four             five
7          four             five
8          four             five
10          two             four
11         four             four
12          two              six
13         four              six
14         four              six
15         four              six
16          two              six
17         four              six
18          two            three
19          two             four
20         four             four

In [81]:
mydata['num-of-doors'].value_counts()

num-of-doors
four    115
two      86
Name: count, dtype: int64

In [82]:
mydata['num-of-doors'].replace('four', 4, inplace=True)
mydata['num-of-doors'].replace('two', 2, inplace=True)

mydata[['num-of-doors','num-of-cylinders']].head(20)

num-of-doors num-of-cylinders
0              2             four
1              2             four
2              2              six
3              4             four
4              4             five
5              2             five
6              4             five
7              4             five
8              4             five
10             2             four
11             4             four
12             2              six
13             4              six
14             4              six
15             4              six
16             2              six
17             4              six
18             2            three
19             2             four
20             4             four

In [83]:
mydata['num-of-cylinders']. value_counts()

num-of-cylinders
four      157
six        24
five       10
two         4
eight       4
three       1
twelve      1
Name: count, dtype: int64

In [85]:
mydata['num-of-cylinders'].replace('four', 4, inplace=True)
mydata['num-of-cylinders'].replace('six', 6, inplace=True)
mydata['num-of-cylinders'].replace('five', 5, inplace=True)
mydata['num-of-cylinders'].replace('two', 2, inplace=True)
mydata['num-of-cylinders'].replace('eight', 8, inplace=True)
mydata['num-of-cylinders'].replace('three', 3, inplace=True)
mydata['num-of-cylinders'].replace('twelve', 12, inplace=True)

mydata[['num-of-doors','num-of-cylinders']].head(20)

num-of-doors  num-of-cylinders
0              2                 4
1              2                 4
2              2                 6
3              4                 4
4              4                 5
5              2                 5
6              4                 5
7              4                 5
8              4                 5
10             2                 4
11             4                 4
12             2                 6
13             4                 6
14             4                 6
15             4                 6
16             2                 6
17             4                 6
18             2                 3
19             2                 4
20             4                 4

In [ ]:
mydata.head()

**One hot-encoding**

In [88]:
#one hot-encoding

conv_fuel = pd.get_dummies(mydata['fuel-type'], dtype=int)

conv_fuel.head()

diesel  gas
0       0    1
1       0    1
2       0    1
3       0    1
4       0    1

In [90]:
mydata_new = pd.concat([mydata, conv_fuel], axis=1)

mydata_new.head()

symboling  normalized-losses         make fuel-type aspiration  \
0          3              122.0  alfa-romero       gas        std   
1          3              122.0  alfa-romero       gas        std   
2          1              122.0  alfa-romero       gas        std   
3          2              164.0         audi       gas        std   
4          2              164.0         audi       gas        std   

   num-of-doors   body-style drive-wheels engine-location  wheel-base  \
0             2  convertible          rwd           front        88.6   
1             2  convertible          rwd           front        88.6   
2             2    hatchback          rwd           front        94.5   
3             4        sedan          fwd           front        99.8   
4             4        sedan          4wd           front        99.4   

     length     width    height  curb-weight engine-type  num-of-cylinders  \
0  0.413433  0.890278 -2.029015         2548        dohc                 4   
1  0.413433  0.890278 -2.029015         2548        dohc                 4   
2  0.449254  0.909722 -0.558319         2823        ohcv                 6   
3  0.529851  0.919444  0.217881         2337         ohc                 4   
4  0.529851  0.922222  0.217881         2824         ohc                 5   

   engine-size fuel-system  bore  stroke  compression-ratio  horsepower  \
0          130        mpfi  3.47    2.68                9.0       111.0   
1          130        mpfi  3.47    2.68                9.0       111.0   
2          152        mpfi  2.68    3.47                9.0       154.0   
3          109        mpfi  3.19    3.40               10.0       102.0   
4          136        mpfi  3.19    3.40                8.0       115.0   

   peak-rpm    price  city_L/100KM  highway_L/100KM  diesel  gas  
0    5000.0  13495.0     11.190476         8.703704       0    1  
1    5000.0  16500.0     11.190476         8.703704       0    1  
2    5000.0  16500.0     12.368421         9.038462       0    1  
3    5500.0  13950.0      9.791667         7.833333       0    1  
4    5500.0  17450.0     13.055556        10.681818       0    1

In [91]:
#one hot-encoding

conv_aspi = pd.get_dummies(mydata['aspiration'], dtype=int)

conv_fuel.head()

diesel  gas
0       0    1
1       0    1
2       0    1
3       0    1
4       0    1

In [93]:
mydata_new1= pd.concat([mydata, conv_aspi], axis=1)

mydata_new1.head()


symboling  normalized-losses         make fuel-type aspiration  \
0          3              122.0  alfa-romero       gas        std   
1          3              122.0  alfa-romero       gas        std   
2          1              122.0  alfa-romero       gas        std   
3          2              164.0         audi       gas        std   
4          2              164.0         audi       gas        std   

   num-of-doors   body-style drive-wheels engine-location  wheel-base  \
0             2  convertible          rwd           front        88.6   
1             2  convertible          rwd           front        88.6   
2             2    hatchback          rwd           front        94.5   
3             4        sedan          fwd           front        99.8   
4             4        sedan          4wd           front        99.4   

     length     width    height  curb-weight engine-type  num-of-cylinders  \
0  0.413433  0.890278 -2.029015         2548        dohc                 4   
1  0.413433  0.890278 -2.029015         2548        dohc                 4   
2  0.449254  0.909722 -0.558319         2823        ohcv                 6   
3  0.529851  0.919444  0.217881         2337         ohc                 4   
4  0.529851  0.922222  0.217881         2824         ohc                 5   

   engine-size fuel-system  bore  stroke  compression-ratio  horsepower  \
0          130        mpfi  3.47    2.68                9.0       111.0   
1          130        mpfi  3.47    2.68                9.0       111.0   
2          152        mpfi  2.68    3.47                9.0       154.0   
3          109        mpfi  3.19    3.40               10.0       102.0   
4          136        mpfi  3.19    3.40                8.0       115.0   

   peak-rpm    price  city_L/100KM  highway_L/100KM  std  turbo  
0    5000.0  13495.0     11.190476         8.703704    1      0  
1    5000.0  16500.0     11.190476         8.703704    1      0  
2    5000.0  16500.0     12.368421         9.038462    1      0  
3    5500.0  13950.0      9.791667         7.833333    1      0  
4    5500.0  17450.0     13.055556        10.681818    1      0

## Data Binning

* **By performing these steps, you:**

  * **Standardize Your Data:** Convert continuous variables into categorical variables, which can be useful for certain types of analysis and modeling.
  * **Enhance Machine Learning Models:** Use the categorized data as features in machine learning models, potentially improving model performance by highlighting meaningful intervals.
  * **Improve Data Visualization:** Make data visualization more intuitive by grouping continuous data into labeled categories, which can simplify the interpretation of charts and graphs.

This process of binning and labeling continuous data is a valuable technique in data preprocessing, enabling more effective data analysis and feature engineering.

In [94]:
mydata['price'].describe()

count      201.000000
mean     13207.129353
std       7947.066342
min       5118.000000
25%       7775.000000
50%      10295.000000
75%      16500.000000
max      45400.000000
Name: price, dtype: float64

In [99]:
price_bins = np.linspace(mydata_new1['price'].min(), mydata_new1['price'].max(), num=4)

print(price_bins)

#Label the bins
price_groups = ['Low Price','Mid Price','High Price']

mydata_new1['price_bins'] = pd.cut(mydata_new1['price'],price_bins, labels = price_groups, include_lowest=True)

mydata_new1[['price','price_bins']].head(10)

[ 5118.         18545.33333333 31972.66666667 45400.        ]


price price_bins
0   13495.0  Low Price
1   16500.0  Low Price
2   16500.0  Low Price
3   13950.0  Low Price
4   17450.0  Low Price
5   15250.0  Low Price
6   17710.0  Low Price
7   18920.0  Mid Price
8   23875.0  Mid Price
10  16430.0  Low Price

In [100]:
mydata_new1['price_bins'].value_counts()

price_bins
Low Price     171
Mid Price      18
High Price     12
Name: count, dtype: int64

In [101]:
mydata_new1.describe()

symboling  normalized-losses  num-of-doors  wheel-base      length  \
count  201.000000          201.00000    201.000000  201.000000  201.000000   
mean     0.840796          122.00000      3.144279   98.797015    0.494045   
std      1.254802           31.99625      0.992008    6.066366    0.183913   
min     -2.000000           65.00000      2.000000   86.600000    0.000000   
25%      0.000000          101.00000      2.000000   94.500000    0.383582   
50%      1.000000          122.00000      4.000000   97.000000    0.479104   
75%      2.000000          137.00000      4.000000  102.400000    0.632836   
max      3.000000          256.00000      4.000000  120.900000    1.000000   

            width        height  curb-weight  num-of-cylinders  engine-size  \
count  201.000000  2.010000e+02   201.000000        201.000000   201.000000   
mean     0.915126  4.949054e-16  2555.666667          4.363184   126.875622   
std      0.029187  1.000000e+00   517.296727          1.059452    41.546834   
min      0.837500 -2.437541e+00  1488.000000          2.000000    61.000000   
25%      0.890278 -7.217300e-01  2169.000000          4.000000    98.000000   
50%      0.909722  1.361755e-01  2414.000000          4.000000   120.000000   
75%      0.925000  7.081124e-01  2926.000000          4.000000   141.000000   
max      1.000000  2.464776e+00  4066.000000         12.000000   326.000000   

             bore      stroke  compression-ratio  horsepower     peak-rpm  \
count  201.000000  201.000000         201.000000  201.000000   201.000000   
mean     3.330692    3.256874          10.164279  103.405534  5117.665368   
std      0.268072    0.316048           4.004965   37.365700   478.113805   
min      2.540000    2.070000           7.000000   48.000000  4150.000000   
25%      3.150000    3.110000           8.600000   70.000000  4800.000000   
50%      3.310000    3.290000           9.000000   95.000000  5125.369458   
75%      3.580000    3.410000           9.400000  116.000000  5500.000000   
max      3.940000    4.170000          23.000000  262.000000  6600.000000   

              price  city_L/100KM  highway_L/100KM         std       turbo  
count    201.000000    201.000000       201.000000  201.000000  201.000000  
mean   13207.129353      9.944145         8.044957    0.820896    0.179104  
std     7947.066342      2.534599         1.840739    0.384397    0.384397  
min     5118.000000      4.795918         4.351852    0.000000    0.000000  
25%     7775.000000      7.833333         6.911765    1.000000    0.000000  
50%    10295.000000      9.791667         7.833333    1.000000    0.000000  
75%    16500.000000     12.368421         9.400000    1.000000    0.000000  
max    45400.000000     18.076923        14.687500    1.000000    1.000000

In [102]:
#Creating Bins for 'normalized-losses'
norm_bins = np.linspace(mydata_new1['normalized-losses'].min(), mydata_new1['normalized-losses'].max(), num=4)

#Printing the bins
print(norm_bins)

#Label the bins
norm_groups = ['Low','Mid','High']

#Assigning Data to the bin
mydata_new1['normalized-losses_bins'] = pd.cut(mydata_new1['normalized-losses'],norm_bins, labels = norm_groups, include_lowest=True)

#Viewing the Binned Data
mydata_new1[['normalized-losses','normalized-losses_bins']].head(10)


[ 65.         128.66666667 192.33333333 256.        ]


normalized-losses normalized-losses_bins
0               122.0                    Low
1               122.0                    Low
2               122.0                    Low
3               164.0                    Mid
4               164.0                    Mid
5               122.0                    Low
6               158.0                    Mid
7               122.0                    Low
8               158.0                    Mid
10              192.0                    Mid

In [103]:
mydata_new1['normalized-losses_bins'].value_counts()

normalized-losses_bins
Low     140
Mid      55
High      6
Name: count, dtype: int64

**Summary from above code**

* **What Are You Doing?**
  * You are categorizing the continuous data in the normalized-losses column into discrete bins labeled 'Low', 'Mid', and 'High'.
  * This process is known as binning, which helps in transforming continuous numerical data into categorical data based on defined ranges.

* **What Do You Learn?**
  * Creating Bins: How to create equally spaced intervals (bins) using numpy.linspace.
  * Labeling Bins: How to assign meaningful labels to these intervals.
  * Using pandas.cut: How to categorize (bin) continuous data into these labeled bins.
  * Inspecting Binned Data: How to verify the binning process by inspecting the results.
* **Why Is This Important?**
  * Simplifies Analysis: Binning converts continuous data into categorical data, making it easier to analyze and interpret.
  * Feature Engineering: Binned data can be used as a feature in machine learning models to improve their performance.
  * Visualization: Categorical data is often easier to visualize and understand in charts and graphs.